src/Analyzer/Passes/RewriteSumFunctionWithSumAndCountPass.cpp

```diff
@@ -0,0 +1,135 @@
+#include <Analyzer/Passes/RewriteSumFunctionWithSumAndCountPass.h>
+
+#include <AggregateFunctions/AggregateFunctionFactory.h>
+#include <Analyzer/InDepthQueryTreeVisitor.h>
+#include <Analyzer/ColumnNode.h>
+#include <Analyzer/ConstantNode.h>
+#include <Analyzer/FunctionNode.h>
+#include <Functions/FunctionFactory.h>
+
+namespace DB
+{
+
+/** Rewrite the following Analyzer to break the function `sum(column +/- literal)` into two individual functions
+ * `sum(column)` and `literal * count(column)`.
+ * sum(column + literal) -> sum(column) + literal * count(column)
+ * sum(literal + column) -> literal * count(column) + sum(column)
+ * sum(column - literal) -> sum(column) - literal * count(column)
+ * sum(literal - column) -> literal * count(column) - sum(column)
+ */
+
+namespace
+{
+
+class RewriteSumFunctionWithSumAndCountVisitor : public InDepthQueryTreeVisitorWithContext<RewriteSumFunctionWithSumAndCountVisitor>
+{
+public:
+    using Base = InDepthQueryTreeVisitorWithContext<RewriteSumFunctionWithSumAndCountVisitor>;
+    using Base::Base;
+
+    void enterImpl(QueryTreeNodePtr & node)
+    {
+        static const std::unordered_set<String> func_supported = {
+            "plus",
+            "minus"
+        };
+
+        auto * function_node = node->as<FunctionNode>();
+        if (!function_node || Poco::toLower(function_node->getFunctionName()) != "sum")
+            return;
+
+        const auto & function_nodes = function_node->getArguments().getNodes();
+        if (function_nodes.size() != 1)
+            return;
+
+        const auto * func_plus_minus_node = function_nodes[0]->as<FunctionNode>();
+        if (!func_plus_minus_node || !func_supported.contains(Poco::toLower(func_plus_minus_node->getFunctionName())))
+            return;
+
+        const auto & func_plus_minus_nodes = func_plus_minus_node->getArguments().getNodes();
+        if (func_plus_minus_nodes.size() != 2)
+            return;
+
+        size_t column_id = func_plus_minus_nodes.size();
```

Comment:
Writing a loop for something that's 2 nodes that must be of different type is super strange (and makes this hard to read)

Reply:
Replace with the following code to decide if column_first/literal_second or literal_first/column_second.

`        size_t column_id;
        if (func_plus_minus_nodes[0]->as<ColumnNode>() && func_plus_minus_nodes[1]->as<ConstantNode>())
            column_id = 0;
        else if (func_plus_minus_nodes[0]->as<ConstantNode>() && func_plus_minus_nodes[1]->as<ColumnNode>())
            column_id = 1;
        else
            return;
`

Reply:
Much better. Thanks!